OSSM 3.0 TP1: OSSM-4983 Configuring Kiali for use with OpenShift Monitoring (#80434)
openshift/openshift-docs: modules/ossm-config-openshift-monitoring-kiali.adoc

// Module included in the following assemblies:
//
// * service-mesh-docs-main/observability/ossm-kiali-assembly.adoc

:_mod-docs-content-type: PROCEDURE
[id="ossm-config-openshift-monitoring-kiali_{context}"]
= Configuring OpenShift Monitoring with Kiali

// TP 1 content. Banner handled separately by different PR handled by Tim O'Keefe. Per Tim, banner should appear across all 3.0 content, nothing needs to be added to this file.
// Possible file name may change
// Possible assembly file may change
// Assemblies, topic map info needs to be worked out still for 3.0.
// Possible content will change.

The following steps show how to integrate the {KialiProduct} with user-workload monitoring.

.Prerequisites

* {SMProductName} is installed.

> * user-workload monitoring is enabled. See link:https://docs.openshift.com/container-platform/4.16/observability/monitoring/enabling-monitoring-for-user-defined-projects.html[Enabling monitoring for user-defined projects].

Reviewer: Since there aren't backticks, the first letter should be capitalized:
```suggestion
* User-workload monitoring is enabled. See link:https://docs.openshift.com/container-platform/4.16/observability/monitoring/enabling-monitoring-for-user-defined-projects.html[Enabling monitoring for user-defined projects].
```

* {KialiProduct} 1.89 is installed.

.Procedure

. Create a `ClusterRoleBinding` resource for Kiali:
+
.Example `ClusterRoleBinding` configuration
[source,yaml]
----
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: kiali-monitoring-rbac
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: cluster-monitoring-view
subjects:
- kind: ServiceAccount
  name: kiali-service-account
  namespace: istio-system
----

. Create a Kiali resource and point it to your Istio instance:
+
.Example Kiali resource configuration
[source,yaml]
----
apiVersion: kiali.io/v1alpha1
kind: Kiali
metadata:
  name: kiali-user-workload-monitoring
  namespace: istio-system
spec:
  external_services:
    prometheus:
      auth:
        type: bearer
        use_kiali_token: true
      thanos_proxy:
        enabled: true
      url: https://thanos-querier.openshift-monitoring.svc.cluster.local:9091
----

> . When the Kiali resource is ready, get the Kiali URL from the Route running the following command:

Reviewer: ```suggestion
. When the Kiali resource is ready, get the Kiali URL from the Route by running the following command:
```

// Module included in the following assemblies:
//
// * service-mesh-docs-main/observability/ossm-kiali-assembly.adoc

:_mod-docs-content-type: PROCEDURE
[id="ossm-config-openshift-monitoring-kiali_{context}"]
= Configuring OpenShift Monitoring with Kiali
:context: ossm-config-openshift-monitoring-kiali

// TP 1 content.
// Possible file name may change
// Possible assembly file may change
// Assemblies, topic map info needs to be worked out still for 3.0.
// Possible content will change.

The following steps show how to integrate the {KialiProduct} with `user-workload monitoring`.

.Prerequisites

* {SMProductName} is installed.

* `user-workload monitoring` is enabled. See "Enabling user-defined projects".

* {KialiProduct} 1.89 is installed.

.Procedure

. Create a `ClusterRoleBinding` resource for Kiali:
+
.Example `ClusterRoleBinding` configuration
[source,yaml]
----
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: kiali-monitoring-rbac
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: cluster-monitoring-view
subjects:
- kind: ServiceAccount
  name: kiali-service-account
  namespace: istio-system
----

> . Create a Kiali resource and point it to your Istio instance:

Reviewer: Not super critical but if this step is telling you to point the resource to the Istio instance, is there a way to signify how or what part of this step achieves that? For instance, if that part of the step is done by specifying the same namespace as the Istio instance, could you add a callout next to the `namespace` parameter in the example and add relevant text in the callout description?

+
.Example Kiali resource configuration
[source,yaml]
----
apiVersion: kiali.io/v1alpha1
kind: Kiali
metadata:
  name: kiali-user-workload-monitoring
  namespace: istio-system
spec:
  external_services:
    prometheus:
      auth:
        type: bearer
        use_kiali_token: true
      thanos_proxy:
        enabled: true
      url: https://thanos-querier.openshift-monitoring.svc.cluster.local:9091
----

> . When the Kiali resource is ready, get the Kiali URL from the `Route` using the following command:

Reviewer: Nit - I would follow one of the options in the [docs guidelines](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#procedure-step-that-introduces-a-command) for introducing a command:
```suggestion
. When the Kiali resource is ready, get the Kiali URL from the `Route` by running the following command:
```

> . When the Kiali resource is ready, get the Kiali URL from the `Route` using the following command:

Reviewer: Similar to my earlier comments, I think `Route` should have a noun following it, or it should have its backticks removed if "Route" or "route" means the same thing in this context

+
[source,terminal]
----

> echo "https://$(oc get routes -n istio-system kiali -o jsonpath='{.spec.host}')"

Reviewer: This command needs an $ in front:
```suggestion
$ echo "https://$(oc get routes -n istio-system kiali -o jsonpath='{.spec.host}')"
```

> echo "https://$(oc get routes -n istio-system kiali -o jsonpath='{.spec.host}')"

Reviewer: Also just checking, disregard if this isn't the case - since there's an oc command seemingly buried in here, does the user need to have the oc CLI installed to do this? If so, I would add "You have installed the OpenShift CLI (`oc`)." as a prereq to this procedure.

The following steps show how to integrate {KialiProduct} with `user-workload monitoring`.

.Prerequisites

* {SMProductName} is installed.

* `user-workload monitoring` is enabled. See "Enabling user-defined projects".

* {KialiProduct} 1.86 is installed.

.Procedure

. Create a `ClusterRoleBinding` resource for Kiali:
+
.Example `ClusterRoleBinding` configuration
[source,yaml]
----
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: kiali-monitoring-rbac
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: cluster-monitoring-view
subjects:
- kind: ServiceAccount
  name: kiali-service-account
  namespace: istio-system
----

. Create a Kiali resource and point it to your Istio instance:
+
.Example Kiali resource configuration
[source,yaml]
----
apiVersion: kiali.io/v1alpha1
kind: Kiali
metadata:
  name: kiali-user-workload-monitoring
  namespace: istio-system
spec:
  external_services:
    prometheus:
      auth:
        type: bearer
        use_kiali_token: true
      thanos_proxy:
        enabled: true
      url: https://thanos-querier.openshift-monitoring.svc.cluster.local:9091
----

. Get the Kiali URL from the `Route`:
+

> [source,terminal]

Reviewer 2: An alternative to getting the URL from the terminal is to go through the console. Do you think this is better?
![AccessKiali](https://github.com/user-attachments/assets/bcf25f62-d98e-4c23-b968-40e9e44380b4)

Author: Hrm...for consistency for 3.0 TP1, let's go through the console. The instructions for Metrics + Kiali have them returning to the console, so to me, it makes sense for users to do the same here, too.

Author: After our Slack discussion, we've decided to keep it as is. Other options for viewing require additional info that is not part of TP 1.

----

> echo "https://$(kubectl get routes -n istio-system kiali -o jsonpath='{.spec.host}')"

Reviewer 2: Sorry this should probably be `oc` for openshift:
```suggestion
echo "https://$(oc get routes -n istio-system kiali -o jsonpath='{.spec.host}')"
```

Reviewer 2: Users also need to wait for the Kiali resource to be ready after creating it in the previous step in order for this command to work. Probably need a note indicating that or else I can provide an `oc wait ...` command to run until this will work?

Author: OK. I revised the next step to say: "When the Kiali resource is ready, get the Kiali URL from the `Route` using the following command:"

//FIX THE STYLE THINGS ONCE DEV HAS REVIEWED/APPROVED THE CONTENT. Content came from different places, so the first priority is ensuring the steps are in the correct order, and are accurate for 3.0 TP1.

The following steps show how to integrate {KialiProduct} with `user-workload monitoring`.

.Prerequisites

* {SMProductName} is installed.

* `user-workload monitoring` is enabled. See "Enabling user-defined projects".

* {KialiProduct} 1.86 is installed.

.Procedure

. Create a `ClusterRoleBinding` for Kiali:
+
.Example `ClusterRoleBinding` configuration
[source,yaml]
----
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: kiali-monitoring-rbac
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: cluster-monitoring-view
subjects:
- kind: ServiceAccount
  name: kiali-service-account
  namespace: istio-system
----

. Find out the revision name of your Istio instance:
+
.Example configuration
[source,terminal]
----
$ kubectl get istiorevisions.sailoperator.io
NAME   READY   STATUS    IN USE   VERSION   AGE
test   True    Healthy   True     v1.21.0   119m9091

> ----

Reviewer 2: The sail-operator documentation is currently out of date. This step isn't necessary anymore.

. Create a Kiali resource and point it to your Istio instance:
+
.Example configuration
[source,yaml]
----
apiVersion: kiali.io/v1alpha1
kind: Kiali
metadata:
  name: kiali-user-workload-monitoring
  namespace: istio-system
spec:
  external_services:

>     istio:
>       config_map_name: istio-test # <1>
>       istio_sidecar_injector_config_map_name: istio-sidecar-injector-test # <2>
>       istiod_deployment_name: istiod-test # <3>
>       url_service_version: 'http://istiod-test.istio-system:15014/version' #<4>

Reviewer 2: Same with this. No need to specify this anymore.
```suggestion
```

    prometheus:
      auth:
        type: bearer
        use_kiali_token: true
      thanos_proxy:
        enabled: true
      url: https://thanos-querier.openshift-monitoring.svc.cluster.local:9091
----

> <1> Replace `istio-test` with your revision name.
> <2> Replace `istio-sidecar-injector-test` with your revision name.
> <3> Replace `istiod-test` with your revision name.
> <4> Replace 'http://istiod-test.istio-system:15014/version` with your revision name.

Reviewer 2: And this. No need to specify these.
```suggestion
```

> // The community docs from which this came and has been taken, https://github.com/istio-ecosystem/sail-operator/tree/main/docs#integrating-with-kiali, do not have validation steps. It is possible there are not validation steps for TP1. Question out to dev 08/21/2024, waiting a response.

Reviewer 2: It actually would be good to add a validation step here! Something that should be added to the community docs as well. The validation step should be

1. Getting the Kiali URL from the `Route`:
    ```
    echo "https://$(kubectl get routes -n istio-system kiali -o jsonpath='{.spec.host}')"
    ```
2. Opening Kiali in your browser by following the URL.
